# Data Cleaning


## Introduction
Greetings from the Machine Learning Team.
This is a qrash course about the most popular Data Analysis python library.



## Getting Started
To use pandas, you'll need to import it after installing from conda or pip.

In [3]:
import pandas as pd

There are two core objects in pandas: the DataFrame and the Series.
### 1. DataFrame
A DataFrame is a table. It contains an array of individual entries, each of which has a certain value. Each entry corresponds to a row (or record) and a column.

<p align="center">
   <img src="https://pandas.pydata.org/docs/_images/01_table_dataframe.svg"/>
</p>

In [8]:
pd.DataFrame({'Tehran': ['Rainy', 'Hazy'], 'Shiraz': ['Partly Sunny', 'Cloudy'], 'Ahwaz': ['Sunshine', 'Sunny']})

Tehran        Shiraz     Ahwaz
0  Rainy  Partly Sunny  Sunshine
1   Hazy        Cloudy     Sunny

The syntax for declaring a new DataFrame is a dictionary whose keys are the column names (Tehran and Shiraz and Ahwaz in this example), and whose values are a list of entries.
to assign values to column labels we can use the below overload:

In [20]:
pd.DataFrame({'Tehran': ['Rainy', 'Hazy'], 'Shiraz': ['Partly Sunny', 'Cloudy'], 'Ahwaz': ['Sunshine', 'Sunny']},index = ['99/08/05','99/08/06'])

Tehran        Shiraz     Ahwaz
99/08/05  Rainy  Partly Sunny  Sunshine
99/08/06   Hazy        Cloudy     Sunny

### 2. Series
A Series, by contrast, is a sequence of data values. If a DataFrame is a table, a Series is a list. And in fact you can create one with nothing more than a list:

In [12]:
pd.Series(['Tehran', 'Shiraz', 'Ahwaz'])

0    Tehran
1    Shiraz
2     Ahwaz
dtype: object

## Reading Data Files
Most of the time, we won't actually be creating our own data by hand. Instead, we'll be working with data that already exists.
<p align="center">
   <img src="https://pandas.pydata.org/docs/_images/02_io_readwrite.svg"/>
</p>
Now, we want to work with a Comma-Seperated Values (csv) file. this is how to read a csv:

In [18]:
covid_dataset = pd.read_csv("COVID-19-Case-Distribution.csv")
covid_dataset.head()

dateRep  day  month  year  cases  deaths countriesAndTerritories geoId  \
0  07/09/2020    7      9  2020     74       2             Afghanistan    AF   
1  06/09/2020    6      9  2020     20       0             Afghanistan    AF   
2  05/09/2020    5      9  2020     16       0             Afghanistan    AF   
3  04/09/2020    4      9  2020     45       1             Afghanistan    AF   
4  03/09/2020    3      9  2020     38       3             Afghanistan    AF   

  countryterritoryCode  popData2019 continentExp  \
0                  AFG   38041757.0         Asia   
1                  AFG   38041757.0         Asia   
2                  AFG   38041757.0         Asia   
3                  AFG   38041757.0         Asia   
4                  AFG   38041757.0         Asia   

   Cumulative_number_for_14_days_of_COVID-19_cases_per_100000  
0                                           1.048847           
1                                           0.854324           
2                                           1.077763           
3                                           1.135594           
4                                           1.272286

We can use the shape attribute to check how large the resulting DataFrame is:

In [19]:
covid_dataset.shape

(41210, 12)

So our Covid-19 DataFrame has 41,210 records split across 12 different columns.

## Dtypes
Data types tell us something about how pandas is storing the data internally. float64 means that it's using a 64-bit floating point number; int64 means a similarly sized integer instead, and so on.
You can use the dtype property to grab the type of a specific column or the whole DataFrame.

In [78]:
print(covid_dataset.deaths.dtype)
print('_________________________')
covid_dataset.dtypes

int64
_________________________


dateRep                                                        object
day                                                             int64
month                                                           int64
year                                                            int64
cases                                                           int64
deaths                                                          int64
countriesAndTerritories                                        object
geoId                                                          object
countryterritoryCode                                           object
popData2019                                                   float64
continentExp                                                   object
Cumulative_number_for_14_days_of_COVID-19_cases_per_100000    float64
dtype: object

## Indexing and Slicing
To retrieve columns of the DataFrame, we use the *iloc* for index-based selection and *loc* for Label-based selection:

In [23]:
covid_dataset.iloc[1:4]

dateRep  day  month  year  cases  deaths countriesAndTerritories geoId  \
1  06/09/2020    6      9  2020     20       0             Afghanistan    AF   
2  05/09/2020    5      9  2020     16       0             Afghanistan    AF   
3  04/09/2020    4      9  2020     45       1             Afghanistan    AF   

  countryterritoryCode  popData2019 continentExp  \
1                  AFG   38041757.0         Asia   
2                  AFG   38041757.0         Asia   
3                  AFG   38041757.0         Asia   

   Cumulative_number_for_14_days_of_COVID-19_cases_per_100000  
1                                           0.854324           
2                                           1.077763           
3                                           1.135594

and to select specific columns like death and countries, we would do:

In [38]:
covid_dataset.iloc[20:25,[6,5]]

countriesAndTerritories  deaths
20             Afghanistan       0
21             Afghanistan       5
22             Afghanistan       7
23             Afghanistan       0
24             Afghanistan       9

In loc operator, it's the data index value, not its position, which matters.

In [36]:
covid_dataset.loc[20:25, ['countriesAndTerritories','deaths']]

countriesAndTerritories  deaths
20             Afghanistan       0
21             Afghanistan       5
22             Afghanistan       7
23             Afghanistan       0
24             Afghanistan       9
25             Afghanistan      10

### Choosing between loc and iloc
*iloc* uses the Python stdlib indexing scheme, where the first element of the range is included and the last one excluded.
So 0:10 will select entries 0,...,9. loc, meanwhile, indexes inclusively. So 0:10 will select entries 0,...,10.

Why the change? Remember that *loc* can index any stdlib type: strings, for example. If we have a DataFrame with index values Apples, ..., Potatoes, ..., and
we want to select "all the alphabetical fruit choices between Apples and Potatoes", then it's a lot more convenient to index *df.loc['Apples':'Potatoes']*
than it is to index something like *df.loc['Apples', 'Potatoet]* (t coming after s in the alphabet).
<br/>

## Selection
So far we've been indexing various strides of data, using structural properties of the DataFrame itself. To do interesting things with the data,
 however, we often need to ask questions based on conditions.
 For example, suppose that we're interested specifically in Cumulative_number_for_14_days_of_COVID-19_cases_per_100000 which are lower than 10:
 We can start by checking if each cumulative number report is lower than 10 or not:

In [47]:
covid_dataset['Cumulative_number_for_14_days_of_COVID-19_cases_per_100000'] < 10

0         True
1         True
2         True
3         True
4         True
         ...  
41205    False
41206    False
41207    False
41208    False
41209    False
Name: Cumulative_number_for_14_days_of_COVID-19_cases_per_100000, Length: 41210, dtype: bool

This operation produced a Series of True/False booleans based on the cumulative number of each record. This result can then be used inside of *loc* to select the relevant data:

In [19]:
covid_dataset.loc[covid_dataset['Cumulative_number_for_14_days_of_COVID-19_cases_per_100000'] < 10]

dateRep  day  month  year  cases  deaths countriesAndTerritories  \
0      07/09/2020    7      9  2020     74       2             Afghanistan   
1      06/09/2020    6      9  2020     20       0             Afghanistan   
2      05/09/2020    5      9  2020     16       0             Afghanistan   
3      04/09/2020    4      9  2020     45       1             Afghanistan   
4      03/09/2020    3      9  2020     38       3             Afghanistan   
...           ...  ...    ...   ...    ...     ...                     ...   
41192  07/04/2020    7      4  2020      0       0                Zimbabwe   
41193  06/04/2020    6      4  2020      0       0                Zimbabwe   
41194  05/04/2020    5      4  2020      0       0                Zimbabwe   
41195  04/04/2020    4      4  2020      1       0                Zimbabwe   
41196  03/04/2020    3      4  2020      0       0                Zimbabwe   

      geoId countryterritoryCode  popData2019 continentExp  \
0        AF                  AFG   38041757.0         Asia   
1        AF                  AFG   38041757.0         Asia   
2        AF                  AFG   38041757.0         Asia   
3        AF                  AFG   38041757.0         Asia   
4        AF                  AFG   38041757.0         Asia   
...     ...                  ...          ...          ...   
41192    ZW                  ZWE   14645473.0       Africa   
41193    ZW                  ZWE   14645473.0       Africa   
41194    ZW                  ZWE   14645473.0       Africa   
41195    ZW                  ZWE   14645473.0       Africa   
41196    ZW                  ZWE   14645473.0       Africa   

       Cumulative_number_for_14_days_of_COVID-19_cases_per_100000  
0                                               1.048847           
1                                               0.854324           
2                                               1.077763           
3                                               1.135594           
4                                               1.272286           
...                                                  ...           
41192                                           0.047796           
41193                                           0.047796           
41194                                           0.047796           
41195                                           0.054624           
41196                                           0.054624           

[23917 rows x 12 columns]

We can use the ampersand (&) to bring the two questions together:

In [53]:
covid_dataset.loc[(covid_dataset.countriesAndTerritories == 'Ukraine') & (covid_dataset['Cumulative_number_for_14_days_of_COVID-19_cases_per_100000'] < 10)]

dateRep  day  month  year  cases  deaths countriesAndTerritories  \
38672  21/04/2020   21      4  2020    261      10                 Ukraine   
38673  20/04/2020   20      4  2020    343       8                 Ukraine   
38674  19/04/2020   19      4  2020    644       8                 Ukraine   
38675  18/04/2020   18      4  2020    301       9                 Ukraine   
38676  17/04/2020   17      4  2020    397       8                 Ukraine   
38677  16/04/2020   16      4  2020    392      10                 Ukraine   
38678  15/04/2020   15      4  2020    270       5                 Ukraine   
38679  14/04/2020   14      4  2020    325      10                 Ukraine   
38680  13/04/2020   13      4  2020    266      10                 Ukraine   
38681  12/04/2020   12      4  2020    308       4                 Ukraine   
38682  11/04/2020   11      4  2020    311      12                 Ukraine   
38683  10/04/2020   10      4  2020    224       5                 Ukraine   
38684  09/04/2020    9      4  2020    206       7                 Ukraine   
38685  08/04/2020    8      4  2020    143       7                 Ukraine   
38686  07/04/2020    7      4  2020     11       1                 Ukraine   
38687  06/04/2020    6      4  2020     83       5                 Ukraine   
38688  05/04/2020    5      4  2020    153       5                 Ukraine   
38689  04/04/2020    4      4  2020    175       5                 Ukraine   
38690  03/04/2020    3      4  2020    103       2                 Ukraine   
38691  02/04/2020    2      4  2020    245       7                 Ukraine   
38692  01/04/2020    1      4  2020     69       2                 Ukraine   
38693  31/03/2020   31      3  2020     62       2                 Ukraine   
38694  30/03/2020   30      3  2020    107       1                 Ukraine   
38695  29/03/2020   29      3  2020     93       3                 Ukraine   
38696  28/03/2020   28      3  2020     62       0                 Ukraine   
38697  27/03/2020   27      3  2020     43       1                 Ukraine   

      geoId countryterritoryCode  popData2019 continentExp  \
38672    UA                  UKR   43993643.0       Europe   
38673    UA                  UKR   43993643.0       Europe   
38674    UA                  UKR   43993643.0       Europe   
38675    UA                  UKR   43993643.0       Europe   
38676    UA                  UKR   43993643.0       Europe   
38677    UA                  UKR   43993643.0       Europe   
38678    UA                  UKR   43993643.0       Europe   
38679    UA                  UKR   43993643.0       Europe   
38680    UA                  UKR   43993643.0       Europe   
38681    UA                  UKR   43993643.0       Europe   
38682    UA                  UKR   43993643.0       Europe   
38683    UA                  UKR   43993643.0       Europe   
38684    UA                  UKR   43993643.0       Europe   
38685    UA                  UKR   43993643.0       Europe   
38686    UA                  UKR   43993643.0       Europe   
38687    UA                  UKR   43993643.0       Europe   
38688    UA                  UKR   43993643.0       Europe   
38689    UA                  UKR   43993643.0       Europe   
38690    UA                  UKR   43993643.0       Europe   
38691    UA                  UKR   43993643.0       Europe   
38692    UA                  UKR   43993643.0       Europe   
38693    UA                  UKR   43993643.0       Europe   
38694    UA                  UKR   43993643.0       Europe   
38695    UA                  UKR   43993643.0       Europe   
38696    UA                  UKR   43993643.0       Europe   
38697    UA                  UKR   43993643.0       Europe   

       Cumulative_number_for_14_days_of_COVID-19_cases_per_100000  
38672                                           9.980987           
38673                                           9.412724           
38674                   

Pandas comes with a few built-in conditional selectors, two of which we will highlight here.

The first is isin. isin is lets you select data whose value "is in" a list of values. For example, here's how we can use it to select records only from Iran or Germany:

In [54]:
covid_dataset.loc[covid_dataset.countriesAndTerritories.isin(['Iran', 'Germany'])]

dateRep  day  month  year  cases  deaths countriesAndTerritories  \
14628  07/09/2020    7      9  2020    814       0                 Germany   
14629  06/09/2020    6      9  2020    988       1                 Germany   
14630  05/09/2020    5      9  2020   2049       5                 Germany   
14631  04/09/2020    4      9  2020    832       0                 Germany   
14632  03/09/2020    3      9  2020   1261       6                 Germany   
...           ...  ...    ...   ...    ...     ...                     ...   
18570  04/01/2020    4      1  2020      0       0                    Iran   
18571  03/01/2020    3      1  2020      0       0                    Iran   
18572  02/01/2020    2      1  2020      0       0                    Iran   
18573  01/01/2020    1      1  2020      0       0                    Iran   
18574  31/12/2019   31     12  2019      0       0                    Iran   

      geoId countryterritoryCode  popData2019 continentExp  \
14628    DE                  DEU   83019213.0       Europe   
14629    DE                  DEU   83019213.0       Europe   
14630    DE                  DEU   83019213.0       Europe   
14631    DE                  DEU   83019213.0       Europe   
14632    DE                  DEU   83019213.0       Europe   
...     ...                  ...          ...          ...   
18570    IR                  IRN   82913893.0         Asia   
18571    IR                  IRN   82913893.0         Asia   
18572    IR                  IRN   82913893.0         Asia   
18573    IR                  IRN   82913893.0         Asia   
18574    IR                  IRN   82913893.0         Asia   

       Cumulative_number_for_14_days_of_COVID-19_cases_per_100000  
14628                                          20.747005           
14629                                          20.622937           
14630                                          20.374802           
14631                                          19.272647           
14632                                          21.073435           
...                                                  ...           
18570                                                NaN           
18571                                                NaN           
18572                                                NaN           
18573                                                NaN           
18574                                                NaN           

[504 rows x 12 columns]

The second is isnull (and its companion notnull). These methods address a frequent problem and let you highlight values which are (or are not) empty (NaN). For example, to filter out records lacking a column in the dataset, here's what we would do:

In [56]:
covid_dataset.loc[covid_dataset['Cumulative_number_for_14_days_of_COVID-19_cases_per_100000'].notnull()]

dateRep  day  month  year  cases  deaths countriesAndTerritories  \
0      07/09/2020    7      9  2020     74       2             Afghanistan   
1      06/09/2020    6      9  2020     20       0             Afghanistan   
2      05/09/2020    5      9  2020     16       0             Afghanistan   
3      04/09/2020    4      9  2020     45       1             Afghanistan   
4      03/09/2020    3      9  2020     38       3             Afghanistan   
...           ...  ...    ...   ...    ...     ...                     ...   
41192  07/04/2020    7      4  2020      0       0                Zimbabwe   
41193  06/04/2020    6      4  2020      0       0                Zimbabwe   
41194  05/04/2020    5      4  2020      0       0                Zimbabwe   
41195  04/04/2020    4      4  2020      1       0                Zimbabwe   
41196  03/04/2020    3      4  2020      0       0                Zimbabwe   

      geoId countryterritoryCode  popData2019 continentExp  \
0        AF                  AFG   38041757.0         Asia   
1        AF                  AFG   38041757.0         Asia   
2        AF                  AFG   38041757.0         Asia   
3        AF                  AFG   38041757.0         Asia   
4        AF                  AFG   38041757.0         Asia   
...     ...                  ...          ...          ...   
41192    ZW                  ZWE   14645473.0       Africa   
41193    ZW                  ZWE   14645473.0       Africa   
41194    ZW                  ZWE   14645473.0       Africa   
41195    ZW                  ZWE   14645473.0       Africa   
41196    ZW                  ZWE   14645473.0       Africa   

       Cumulative_number_for_14_days_of_COVID-19_cases_per_100000  
0                                               1.048847           
1                                               0.854324           
2                                               1.077763           
3                                               1.135594           
4                                               1.272286           
...                                                  ...           
41192                                           0.047796           
41193                                           0.047796           
41194                                           0.047796           
41195                                           0.054624           
41196                                           0.054624           

[38429 rows x 12 columns]

### Assigning Value
Going the other way, assigning data to a DataFrame is easy. You can assign either a constant value:

In [66]:
covid_dataset['deaths'] = 274
covid_dataset['deaths']

0        274
1        274
2        274
3        274
4        274
        ... 
41205    274
41206    274
41207    274
41208    274
41209    274
Name: deaths, Length: 41210, dtype: int64

Or with an iterable of values:

In [59]:
covid_dataset['index_backward'] = range(covid_dataset.shape[0], 0, -1)
covid_dataset['index_backward']

0        41210
1        41209
2        41208
3        41207
4        41206
         ...  
41205        5
41206        4
41207        3
41208        2
41209        1
Name: index_backward, Length: 41210, dtype: int32

## Renaming
Oftentimes data will come to us with column names, index names, or other naming conventions that we are not satisfied with.
 In that case, you'll learn how to use pandas functions to change the names of the offending entries to something better.

The first function we'll introduce here is rename(), which lets you change index names and/or column names. For example, to change the countriesAndTerritories column in our dataset to location, we would do:

In [135]:
covid_dataset = covid_dataset.rename(columns={'countriesAndTerritories': 'location'})
covid_dataset = covid_dataset.rename(columns={'Cumulative_number_for_14_days_of_COVID-19_cases_per_100000': 'recent_cases_per_100000'})

You'll probably rename columns very often, but rename index values very rarely. For that, set_index() is usually more convenient.


## Missing Data
Entries missing values are given the value NaN, short for "Not a Number". For technical reasons these NaN values are always of the float64 dtype.

Pandas provides some methods specific to missing data. To select NaN entries you can use pd.isnull() (or its companion pd.notnull()). This is meant to be used thusly:

In [81]:
covid_dataset[pd.isnull(covid_dataset.location)]

Empty DataFrame
Columns: [dateRep, day, month, year, cases, deaths, location, geoId, countryterritoryCode, popData2019, continentExp, Cumulative_number_for_14_days_of_COVID-19_cases_per_100000]
Index: []

But for

In [83]:
covid_dataset[pd.isnull(covid_dataset.recent_cases_per_100000)]

dateRep  day  month  year  cases  deaths     location geoId  \
229    12/01/2020   12      1  2020      0       0  Afghanistan    AF   
230    11/01/2020   11      1  2020      0       0  Afghanistan    AF   
231    10/01/2020   10      1  2020      0       0  Afghanistan    AF   
232    09/01/2020    9      1  2020      0       0  Afghanistan    AF   
233    08/01/2020    8      1  2020      0       0  Afghanistan    AF   
...           ...  ...    ...   ...    ...     ...          ...   ...   
41205  25/03/2020   25      3  2020      0       0     Zimbabwe    ZW   
41206  24/03/2020   24      3  2020      0       1     Zimbabwe    ZW   
41207  23/03/2020   23      3  2020      0       0     Zimbabwe    ZW   
41208  22/03/2020   22      3  2020      1       0     Zimbabwe    ZW   
41209  21/03/2020   21      3  2020      1       0     Zimbabwe    ZW   

      countryterritoryCode  popData2019 continentExp  recent_cases_per_100000  
229                    AFG   38041757.0         Asia                      NaN  
230                    AFG   38041757.0         Asia                      NaN  
231                    AFG   38041757.0         Asia                      NaN  
232                    AFG   38041757.0         Asia                      NaN  
233                    AFG   38041757.0         Asia                      NaN  
...                    ...          ...          ...                      ...  
41205                  ZWE   14645473.0       Africa                      NaN  
41206                  ZWE   14645473.0       Africa                      NaN  
41207                  ZWE   14645473.0       Africa                      NaN  
41208                  ZWE   14645473.0       Africa                      NaN  
41209                  ZWE   14645473.0       Africa                      NaN  

[2781 rows x 12 columns]

and you can deal with these data with *fillna* method which provides a few different strategies for mitigating such data.  For example, we can simply replace each NaN with an "Unknown":

In [84]:
covid_dataset.recent_cases_per_100000.fillna('Unknown')

0         1.04885
1        0.854324
2         1.07776
3         1.13559
4         1.27229
           ...   
41205     Unknown
41206     Unknown
41207     Unknown
41208     Unknown
41209     Unknown
Name: recent_cases_per_100000, Length: 41210, dtype: object

Alternatively, we may have a non-null value that we would like to replace. For example, suppose that since this dataset was published and a bug fix has reported accordingly, One way to reflect this in the dataset is using the replace() method:

In [119]:
covid_dataset.loc[covid_dataset.countryterritoryCode == 'AFG'].cases.replace(74, 70)

0      70
1      20
2      16
3      45
4      38
       ..
237     0
238     0
239     0
240     0
241     0
Name: cases, Length: 242, dtype: int64

In [121]:
covid_dataset

dateRep  day  month  year  cases  deaths countriesAndTerritories  \
0      07/09/2020    7      9  2020     74       2             Afghanistan   
1      06/09/2020    6      9  2020     20       0             Afghanistan   
2      05/09/2020    5      9  2020     16       0             Afghanistan   
3      04/09/2020    4      9  2020     45       1             Afghanistan   
4      03/09/2020    3      9  2020     38       3             Afghanistan   
...           ...  ...    ...   ...    ...     ...                     ...   
41205  25/03/2020   25      3  2020      0       0                Zimbabwe   
41206  24/03/2020   24      3  2020      0       1                Zimbabwe   
41207  23/03/2020   23      3  2020      0       0                Zimbabwe   
41208  22/03/2020   22      3  2020      1       0                Zimbabwe   
41209  21/03/2020   21      3  2020      1       0                Zimbabwe   

      geoId countryterritoryCode  popData2019 continentExp  \
0        AF                  AFG   38041757.0         Asia   
1        AF                  AFG   38041757.0         Asia   
2        AF                  AFG   38041757.0         Asia   
3        AF                  AFG   38041757.0         Asia   
4        AF                  AFG   38041757.0         Asia   
...     ...                  ...          ...          ...   
41205    ZW                  ZWE   14645473.0       Africa   
41206    ZW                  ZWE   14645473.0       Africa   
41207    ZW                  ZWE   14645473.0       Africa   
41208    ZW                  ZWE   14645473.0       Africa   
41209    ZW                  ZWE   14645473.0       Africa   

       Cumulative_number_for_14_days_of_COVID-19_cases_per_100000  
0                                               1.048847           
1                                               0.854324           
2                                               1.077763           
3                                               1.135594           
4                                               1.272286           
...                                                  ...           
41205                                                NaN           
41206                                                NaN           
41207                                                NaN           
41208                                                NaN           
41209                                                NaN           

[41210 rows x 12 columns]

## Summary Functions
### describe
Pandas provides many simple "summary functions" (not an official name) which restructure the data in some useful way. For example, consider the describe() method:
#### 1. Numerical

In [122]:
covid_dataset.popData2019.describe()

count    4.114600e+04
mean     4.324698e+07
std      1.588085e+08
min      8.150000e+02
25%      1.394969e+06
50%      8.519373e+06
75%      2.916192e+07
max      1.433784e+09
Name: popData2019, dtype: float64

#### 2. String

In [124]:
covid_dataset.continentExp.describe()

count      41210
unique         6
top       Europe
freq       11804
Name: continentExp, dtype: object

### mean()
If you want to get some particular simple summary statistic about a column in a DataFrame or a Series, there is usually a helpful pandas function that makes it happen.

For example, to see the mean of the points allotted,we can use the mean() method:

In [125]:
covid_dataset.cases.mean()


658.8400145595729

### unique()
To see a list of unique values we can use the unique() method:

In [126]:
covid_dataset.dateRep.unique()

array(['07/09/2020', '06/09/2020', '05/09/2020', '04/09/2020',
       '03/09/2020', '02/09/2020', '01/09/2020', '31/08/2020',
       '30/08/2020', '29/08/2020', '28/08/2020', '27/08/2020',
       '26/08/2020', '25/08/2020', '24/08/2020', '23/08/2020',
       '22/08/2020', '21/08/2020', '20/08/2020', '19/08/2020',
       '18/08/2020', '17/08/2020', '16/08/2020', '15/08/2020',
       '14/08/2020', '13/08/2020', '12/08/2020', '11/08/2020',
       '10/08/2020', '09/08/2020', '08/08/2020', '07/08/2020',
       '06/08/2020', '05/08/2020', '04/08/2020', '03/08/2020',
       '02/08/2020', '01/08/2020', '31/07/2020', '30/07/2020',
       '29/07/2020', '28/07/2020', '27/07/2020', '26/07/2020',
       '25/07/2020', '24/07/2020', '23/07/2020', '22/07/2020',
       '21/07/2020', '20/07/2020', '19/07/2020', '18/07/2020',
       '17/07/2020', '16/07/2020', '15/07/2020', '14/07/2020',
       '13/07/2020', '12/07/2020', '11/07/2020', '10/07/2020',
       '09/07/2020', '08/07/2020', '07/07/2020', '06/07

### value_counts()
To see a list of unique values and how often they occur in the dataset, we can use the value_counts() method:

## Maps
In data science we often have a need for creating new representations from existing data, or for transforming data from the format it is in now to the format that we want it to be in later. Maps are what handle this work, making them extremely important for getting your work done!

There are two mapping methods that you will use often.

### map()
suppose that we wanted to remean the deaths to 0. We can do this as follows:

In [128]:
deaths_mean = covid_dataset.deaths.mean()
covid_dataset.deaths.map(lambda p: p - deaths_mean)

0       -19.578646
1       -21.578646
2       -21.578646
3       -20.578646
4       -18.578646
           ...    
41205   -21.578646
41206   -20.578646
41207   -21.578646
41208   -21.578646
41209   -21.578646
Name: deaths, Length: 41210, dtype: float64

The function you pass to map() should expect a single value from the Series (a point value, in the above example), and return a transformed version of that value. map() returns a new Series where all the values have been transformed by your function.

### apply()
 is the equivalent method if we want to transform a whole DataFrame by calling a custom method on each row.

In [129]:
def remean(row):
    row.deaths = row.deaths - deaths_mean
    return row
covid_dataset.apply(remean, axis='columns')

dateRep  day  month  year  cases     deaths countriesAndTerritories  \
0      07/09/2020    7      9  2020     74 -19.578646             Afghanistan   
1      06/09/2020    6      9  2020     20 -21.578646             Afghanistan   
2      05/09/2020    5      9  2020     16 -21.578646             Afghanistan   
3      04/09/2020    4      9  2020     45 -20.578646             Afghanistan   
4      03/09/2020    3      9  2020     38 -18.578646             Afghanistan   
...           ...  ...    ...   ...    ...        ...                     ...   
41205  25/03/2020   25      3  2020      0 -21.578646                Zimbabwe   
41206  24/03/2020   24      3  2020      0 -20.578646                Zimbabwe   
41207  23/03/2020   23      3  2020      0 -21.578646                Zimbabwe   
41208  22/03/2020   22      3  2020      1 -21.578646                Zimbabwe   
41209  21/03/2020   21      3  2020      1 -21.578646                Zimbabwe   

      geoId countryterritoryCode  popData2019 continentExp  \
0        AF                  AFG   38041757.0         Asia   
1        AF                  AFG   38041757.0         Asia   
2        AF                  AFG   38041757.0         Asia   
3        AF                  AFG   38041757.0         Asia   
4        AF                  AFG   38041757.0         Asia   
...     ...                  ...          ...          ...   
41205    ZW                  ZWE   14645473.0       Africa   
41206    ZW                  ZWE   14645473.0       Africa   
41207    ZW                  ZWE   14645473.0       Africa   
41208    ZW                  ZWE   14645473.0       Africa   
41209    ZW                  ZWE   14645473.0       Africa   

       Cumulative_number_for_14_days_of_COVID-19_cases_per_100000  
0                                               1.048847           
1                                               0.854324           
2                                               1.077763           
3                                               1.135594           
4                                               1.272286           
...                                                  ...           
41205                                                NaN           
41206                                                NaN           
41207                                                NaN           
41208                                                NaN           
41209                                                NaN           

[41210 rows x 12 columns]

In this code we are performing an operation between a lot of values on the left-hand side (everything in the Series) and a single value on the right-hand side (the mean value). Pandas looks at this expression and figures out that we must mean to subtract that mean value from every value in the dataset.

Pandas will also understand what to do if we perform these operations between Series of equal length. For example, an easy way of combining location and cases information in the dataset would be to do the following:

In [137]:
covid_dataset.location + " - " + covid_dataset.cases

0        Afghanistan - 0        74\n1        20\n2     ...
1        Afghanistan - 0        74\n1        20\n2     ...
2        Afghanistan - 0        74\n1        20\n2     ...
3        Afghanistan - 0        74\n1        20\n2     ...
4        Afghanistan - 0        74\n1        20\n2     ...
                               ...                        
41205    Zimbabwe - 0        74\n1        20\n2        ...
41206    Zimbabwe - 0        74\n1        20\n2        ...
41207    Zimbabwe - 0        74\n1        20\n2        ...
41208    Zimbabwe - 0        74\n1        20\n2        ...
41209    Zimbabwe - 0        74\n1        20\n2        ...
Name: location, Length: 41210, dtype: object

## Combining
When performing operations on a dataset, we will sometimes need to combine different DataFrames and/or Series in non-trivial ways.
 Pandas has three core methods for doing this. In order of increasing complexity, these are concat(), join(), and merge().
 Most of what merge() can do can also be done more simply with join(), so we will omit it and focus on the first two functions here.

<p align="center">
   <img src="https://pandas.pydata.org/docs/_images/08_concat_row.svg"/>
</p>

### Concat
The simplest combining method is concat().
 Given a list of elements, this function will smush those elements together along an axis.

Along rows:

In [16]:
df1 = pd.DataFrame({'A': ['A0', 'A1', 'A2', 'A3'],
                    'B': ['B0', 'B1', 'B2', 'B3'],
                    'C': ['C0', 'C1', 'C2', 'C3'],
                    'D': ['D0', 'D1', 'D2', 'D3']},
                   index=[0, 1, 2, 3])
df2 = pd.DataFrame({'A': ['A4', 'A5', 'A6', 'A7'],
                    'B': ['B4', 'B5', 'B6', 'B7'],
                    'C': ['C4', 'C5', 'C6', 'C7'],
                    'D': ['D4', 'D5', 'D6', 'D7']},
                   index=[4, 5, 6, 7])
df3 = pd.DataFrame({'A': ['A8', 'A9', 'A10', 'A11'],
                    'B': ['B8', 'B9', 'B10', 'B11'],
                    'C': ['C8', 'C9', 'C10', 'C11'],
                    'D': ['D8', 'D9', 'D10', 'D11']},
                   index=[8, 9, 10, 11])
pd.concat([df1, df2, df3], axis='rows')

A    B    C    D
0    A0   B0   C0   D0
1    A1   B1   C1   D1
2    A2   B2   C2   D2
3    A3   B3   C3   D3
4    A4   B4   C4   D4
5    A5   B5   C5   D5
6    A6   B6   C6   D6
7    A7   B7   C7   D7
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11

Along columns: ???

In [6]:
pd.concat([df1, df2, df3], axis='columns')

A    B    C    D    A    B    C    D    A    B    C    D
0    A0   B0   C0   D0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1    A1   B1   C1   D1  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2    A2   B2   C2   D2  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
3    A3   B3   C3   D3  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN
4   NaN  NaN  NaN  NaN   A4   B4   C4   D4  NaN  NaN  NaN  NaN
5   NaN  NaN  NaN  NaN   A5   B5   C5   D5  NaN  NaN  NaN  NaN
6   NaN  NaN  NaN  NaN   A6   B6   C6   D6  NaN  NaN  NaN  NaN
7   NaN  NaN  NaN  NaN   A7   B7   C7   D7  NaN  NaN  NaN  NaN
8   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   A8   B8   C8   D8
9   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   A9   B9   C9   D9
10  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  A10  B10  C10  D10
11  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  A11  B11  C11  D11

mind the indexes

In [17]:
df2.index = df1.index
pd.concat([df1, df2], axis='columns')

A   B   C   D   A   B   C   D
0  A0  B0  C0  D0  A4  B4  C4  D4
1  A1  B1  C1  D1  A5  B5  C5  D5
2  A2  B2  C2  D2  A6  B6  C6  D6
3  A3  B3  C3  D3  A7  B7  C7  D7

### Join
The middlemost combiner in terms of complexity is join().
 join() lets you combine different DataFrame objects which have an index in common.

In [239]:
df2.A = df1.A # Making a common column
left = df1.set_index(['A'])
right = df2.set_index(['A'])
left.join(right, lsuffix='_from_df1', rsuffix='_from_df2')

B_from_df1 C_from_df1 D_from_df1 B_from_df2 C_from_df2 D_from_df2
A                                                                   
A0         B0         C0         D0         B4         C4         D4
A1         B1         C1         D1         B5         C5         D5
A2         B2         C2         D2         B6         C6         D6
A3         B3         C3         D3         B7         C7         D7